src/main/java/org/fungover/thunder/ClientHandler.java

```diff
@@ -1,41 +1,33 @@
 package org.fungover.thunder;
 
-import java.io.IOException;
-import java.net.ServerSocket;
 import java.net.Socket;
-import java.util.ArrayList;
-import java.util.Collections;
 import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
 
 public class ClientHandler {
     private final List<Socket> clients;
     private final PackageReader packageReader;
 
     public ClientHandler() {
-        clients = Collections.synchronizedList(new ArrayList<>());
+        clients = new CopyOnWriteArrayList<>();
         packageReader = new PackageReader();
     }
 
-    public void handleConnections(ServerSocket serverSocket) throws IOException {
-        while (!serverSocket.isClosed()) {
-            Socket connection = null;
-            try {
-                connection = serverSocket.accept();
-            } catch (IOException e) {
-                e.printStackTrace();
-            }
-            if (connection != null) {
-                if (packageReader.isValidConnection(connection)) {
-                    clients.add(connection);
-                }else{
-                    connection.close();
+    public void handleConnection(Socket clientSocket) {
+        try {
+            while (!clientSocket.isClosed()) {
+                System.out.println("New client: " + clientSocket.getInetAddress().getHostName());
+                if (packageReader.isValidConnection(clientSocket)) {
+                    clients.add(clientSocket);
+                } else {
+                    clientSocket.close();
                 }
             }
```

Comment:
The while-loop disconnects the client after connecting because isValidConnection() will return false the second loop.
We need the while-loop to continue listening for more packages from the client, up until the client sends the disconnect package.
So maybe we need some other check where we close the socket if it is a duplicate connection and not only if isValidConnection() fails? Or since one connection can only send one connect and recieve one connack response, limit the times one client can check the connection?

Reply:
> So maybe we need some other check where we close the socket if it is a duplicate connection and not only if isValidConnection() fails? Or since one connection can only send one connect and recieve one connack response, limit the times one client can check the connection?

There is an issue for the disconnection handling. So that feature will come later.

> The while-loop disconnects the client after connecting because isValidConnection() will return false the second loop.
> We need the while-loop to continue listening for more packages from the client, up until the client sends the disconnect package.

#16 is the only package that is being sent right now. I think packagehandler gets modified when other functions are added. But I can remove the while loop again if you want and just close the socket after adding it? I brought the loop back this time because I consideret it possible that we need a while loop in ClientHandler class after all 😄

Reply:
```suggestion
   public void handleConnection(Socket clientSocket) {
        try {
            while (!clientSocket.isClosed()) {
                System.out.println("New client: " + clientSocket.getInetAddress().getHostName());
                if (packageReader.isValidConnection(clientSocket)) {
                    clients.add(clientSocket);
                }
            }
            clientSocket.close();
    } catch(
    Exception e)
    {
        e.printStackTrace();
    }
}
```

This might be an option?

Reply:
```suggestion
public void handleConnection(Socket clientSocket) {
        try {
            if (packageReader.isValidConnection(clientSocket)) {
                System.out.println("New client: " + clientSocket.getInetAddress().getHostName());
                clients.add(clientSocket);
            } else {
                //Disconnect client/close socket
            }
            while (!clientSocket.isClosed()) {
                //Logic to read from or disconnect client.
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
```

Is it something like this you would suggest? @Win-ther

Reply:
> There is an issue for the disconnection handling. So that feature will come later.
>
Yes I know, I have assigned myself to the issue and I'm working on it 



> #16 is the only package that is being sent right now. I think packagehandler gets modified when other functions are added. But I can remove the while loop again if you want and just close the socket after adding it? I brought the loop back this time because I consideret it possible that we need a while loop in ClientHandler class after all 😄
> 
> Could you add some code? Would be easier for me to see what you suggest 👍

No I think we need a loop in clienthandler. The problem is that the clientsocket gets closed instantly after the CONNECT message is recieved and CONNACK is sent. 
```suggestion
    public void handleConnection(Socket clientSocket) {
    try {
                if (packageReader.isValidConnection(clientSocket)) {
                    System.out.println("New client: " + clientSocket.getInetAddress().getHostName());
                    clients.add(clientSocket);
                } else{
                    clientSocket.close();
                }
            while (!clientSocket.isClosed()) {
                //ToDo: Listen for packets ex. disconnect or ping
            }
```
Here is a quick suggestion, since we only need to connect once. The isValidConnection()-call can be outside the loop, and have the loop listen for future packets that will be implemented in other Issues

Reply:
> Is it something like this you would suggest? @Win-ther

Yes that looks good, I think I just commented the same as you suggested. Sorry it took some time, I had some other stuff to do😅😬

Reply:
Great! I did not understand at first, but I am happy we ended up with the same solution 😃 Don't worry, the conversation was set to resolved so I wasnt sure you would be noticed of my response. Thanks for feedback, I will update the code.